### PR TITLE
Coin Selection Better Coverage + Fix Fee Calculation

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

# Overview

- [x] I have improved coin selection unit tests by also specifying and checking other parts of a coin selection result (i.e. outputs and change)
- [x] I have added some edge-case unit tests to test coin overflow in some cases
- [x] I have fixed a wrong strict inequality sign (caught by updated tests)
- [x] I have fixed a bug in the fee estimation were we were counting the same bytes twice.

# Comments

@piotr-iohk This is the best I can do in terms of coverage in a reasonable amount of time. There's still one particular overflow case that is hard to instrument in practice (impossible to meet :thinking: ?) but that is there in the code as a safe "gatekeeper" I suppose.

### CODE REVIEW

File: src/Cardano/Wallet/CoinSelection/Fee.hs [hunk 1]
```diff
@@ -421,10 +419,7 @@ estimateFee policy (CoinSelection inps outs chngs) =
     --  | word64                      -- 1|2|3|5|9
     sizeOfTxOut :: TxOut -> Int
     sizeOfTxOut (TxOut (Address bytes) c) =
-        6
-        + BS.length bytes
-        + sizeOf (CBOR.encodeWord32 $ crc32 bytes)
-        + sizeOfCoin c
+        1 + BS.length bytes + sizeOfCoin c
```
Comment (KtorZ) [inline src/Cardano/Wallet/CoinSelection/Fee.hs:422]: So, turns out those 5 bytes + the crc are already part of the `bytes` payload! So we don't need to count them twice!

File: test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs [hunk 2]
```diff
@@ -377,15 +408,17 @@ propFeeEstimation (ShowFmt sel, InfiniteList chngAddrs _) =
...
         -- We always go for the higher bound for change address payload's size,
         -- so, we may end up with up to 4 extra bytes per change address in our
         -- estimation.
         margin = 4 * fromIntegral (length $ CS.change sel)
         realFeeSup = ceiling (a + b*(fromIntegral size + margin))
         realFeeInf = ceiling (a + b*(fromIntegral size))
     in
-        property (calcFee >= realFeeInf && calcFee <= realFeeSup)
+        (calcFee >= realFeeInf && calcFee <= realFeeSup, encodedTx)
+            === (True, encodedTx)
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:421]: No change in the test logic here, but we'll now display the full transaction binary representation upon failure which is quite handy for debugging!

File: test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs [hunk 3]
```diff
@@ -524,8 +557,8 @@ instance Arbitrary Address where
 instance {-# OVERLAPS #-} Arbitrary (Network -> Address) where
     shrink _ = []
     arbitrary = do
-        mainnetA <- genAddress (39, 43)
-        testnetA <- genAddress (46, 50)
+        mainnetA <- genAddress (33, 33)
+        testnetA <- genAddress (40, 40)
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:561]: This ought to be only the internal payload of the address. So, same confusion as within the fee calculation but, on the generation this time. The error was hard to catch because this one actually compensate for the previous one in the implementation. So, only in rare cases (1 tests out of 33k!), the double-counting would not be compensated by the wrong generation here and trigger an error!

File: test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs [hunk 4]
```diff
@@ -299,7 +300,7 @@ spec = do
 
     describe "Fee Estimation properties" $ do
         it "Estimated fee is the same as taken by encodeSignedTx"
-            (withMaxSuccess 1000 $ property propFeeEstimation)
+            (withMaxSuccess 2500 $ property propFeeEstimation)
```
Comment (piotr-iohk) [inline test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:333]: more the merrier :) :+1:

**Review by piotr-iohk — APPROVED** [review]
lgtm :+1:
